compiler/daml-lf-verify/src/DA/Daml/LF/Verify/Context.hs

```diff
@@ -0,0 +1,138 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TemplateHaskell #-}
+
+-- | Contexts for DAML LF static verification
+module DA.Daml.LF.Verify.Context
+  ( Delta
+  , MonadDelta, devars, devals
+  , UpdateSet(..)
+  , UpdCreate(..), usCre, usArc, usCho
+  , UpdArchive(..)
+  , UpdChoice(..)
+  , runDelta
+  , emptyDelta
+  , introDelta, extVarDelta
+  , lookupDExprVar, lookupDVal, lookupDChoice
+  , concatDelta
+  , emptyUpdateSet
+  , concatUpdateSet
+  ) where
+
+import Control.Lens hiding (Context)
+import Control.Monad.Error.Class (MonadError (..))
+import Control.Monad.Reader
+import Data.HashMap.Strict (HashMap, union, empty)
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Error
+
+data UpdCreate = UpdCreate
+  { _creTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _creField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+data UpdArchive = UpdArchive
+  { _arcTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _arcField :: ![(FieldName, Expr)]
+    -- ^ The fields to be verified, together with their value.
+  }
+data UpdChoice = UpdChoice
+  { _choTemp  :: !(Qualified TypeConName)
+    -- ^ Qualified type constructor corresponding to the contract template.
+  , _choName  :: !ChoiceName
+    -- ^ The name of the choice.
+  }
+
+-- | The List of updates being performed
+data UpdateSet = UpdateSet
+  { _usCre :: ![UpdCreate]
+    -- ^ The list of create updates.
+  , _usArc :: ![UpdArchive]
+    -- ^ The list of archive updates.
+  , _usCho :: ![UpdChoice]
+    -- ^ The list of choice updates.
+  }
+
+makeLenses ''UpdateSet
```

Comment:
We generally avoid TemplateHaskell for the most part. Part of that is just a code style guideline since we generally try write fairly simple Haskell code but for TemplateHaskell there is also a technical resaon: We use a statically linked GHC (I won’t go into the reasons here but I’m happy to expand on that in another context) which makes TemplateHaskell very slow. So if you really do want to use template haskell. You are better off putting all the template Haskell code into a separate Bazel target that changes rarely (Bazel caches at the granularity of individual targets). Otherwise, you will suffer from long compile times while working on this.

However, we also mostly avoid lenses. This is not quite as strict and there is not really a technical reason for this but I’m not aware of any module that uses them as pervasively as you use them here.

So overall, I think you’re better off removing the TH lens generation, switch away from using lenses by default for everything and potentially define a few lenses by hand for the things where they do provide a significant benefit.